ros_packages/voice_assistant/voice_assistant/audio_player.py

```diff
@@ -0,0 +1,193 @@
+            "play_audio_from_speech", 
+            self.play_audio_from_speech, 
+            callback_group=play_audio_callback_group)
+        
+        # clears the playback queue
+        self.clear_playback_queue_servie = self.create_service(
+            ClearPlaybackQueue,
+            "clear_playback_queue", 
+            self.clear_playback_queue, 
+            callback_group=clear_callback_group)
+
+        # service for playing back audio files (must be wav-format)
+        self.play_audio_from_file_service = self.create_service(
+            PlayAudioFromFile,
+            "play_audio_from_file", 
+            self.play_audio_from_file, 
+            callback_group=play_audio_callback_group)
+        
+        self.get_logger().info('Now running AUDIO PLAYER')
+
+
+
+    def counter_next(self) -> int:
+
+        with self.counter_lock:
+            value = self.counter
+            self.counter += 1
+            return value
+
+
+    
+    def play_audio_from_file(self, request: PlayAudioFromFile.Request, response: PlayAudioFromFile.Response) -> PlayAudioFromFile.Response:
+
+        order = self.counter_next()
+
+        with wave.open(request.filepath, 'rb') as wf:
+
+            encoding = AudioEncoding(
+                wf.getsampwidth(),
+                wf.getnchannels(),
+                wf.getframerate())
+            
+            frames_per_chunk = encoding.frames_per_second // CHUNKS_PER_SECOND
+            
+            data = []
+            while True:
+                chunk = wf.readframes(frames_per_chunk)
+                data.append(chunk)
+                if len(chunk) < frames_per_chunk: break
+
+        playback_item = PlaybackItem(data, encoding, 0.0, order)
+        self.playback_queue.put(playback_item, True)
+
+        if request.join: playback_item.finished_playing.wait()
+        return response
+
+
+
+    def play_audio_from_speech(self, request: PlayAudioFromSpeech.Request, response: PlayAudioFromSpeech.Response) -> PlayAudioFromSpeech.Response:
+
+        order = self.counter_next()
+
+        data = public_voice_client.text_to_speech(request.speech, request.gender, request.language)
+
+        playback_item = PlaybackItem(data, SPEECH_ENCODING, 0.2, order)
+        self.playback_queue.put(playback_item, True)
+
+        if request.join: playback_item.finished_playing.wait()
+        return response
+
+
+
+    def clear_playback_queue(self, _: PlayAudioFromFile.Request, response: PlayAudioFromFile.Response) -> PlayAudioFromFile.Response:
+
+        with PlaybackItem.clear_threshold_lock: 
+            with self.counter_lock:
+                PlaybackItem.clear_threshold = self.counter
+
+        return response
+
+
+
+def main(args=None):
+
+    playback_queue: Queue[PlaybackItem] = Queue()
+
+    def audio_loop(playback_queue: Queue[PlaybackItem]) ->  None: 
+        while True: playback_queue.get(True).play()
+
+    Thread(target=audio_loop, args=(playback_queue,), daemon=True).start()
```

Comment:
Shouldn't this be part of the AudioPlayerNode since it is the core component of playing the audio?

Also couldn't this loop be used to pop elements from the queue and play them? I feel a queue is the perfect datatype for this type of task, instead of letting each item handle the responsibility of whether it should be played or not. As of now, it seems like items are never removed from the queue. That means the longer pib will be running, the more this queue will fill up with data that is not needed anymore.

Reply:
the `Node` class is mainly meant to provide callbacks and entities for ROS-communication. ROS does not provide any means of running a background task over the whole lifecycle of a node (unless you use a timer, which may incur latency penalties), so this task is run as a daemon-thread. the `get` method of the `Queue` class automatically pops the last element.